<a href="https://colab.research.google.com/github/harshnandwana/Machine-learning-projects/blob/main/Suicide_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!wget --header 'Host: storage.googleapis.com' --user-agent 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:101.0) Gecko/20100101 Firefox/101.0' --header 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8' --header 'Accept-Language: en-US,en;q=0.5' --referer 'https://www.kaggle.com/' --header 'DNT: 1' --header 'Alt-Used: storage.googleapis.com' --header 'Upgrade-Insecure-Requests: 1' --header 'Sec-Fetch-Dest: document' --header 'Sec-Fetch-Mode: navigate' --header 'Sec-Fetch-Site: cross-site' --header 'Sec-Fetch-User: ?1' 'https://storage.googleapis.com/kaggle-data-sets/1312443/2186313/bundle/archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20220617%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20220617T080425Z&X-Goog-Expires=259199&X-Goog-SignedHeaders=host&X-Goog-Signature=2c284e453351f7bdecda66f4696b76039d2ed56afbab1d23e570ba86533e41acdb5c08d835a8174de45406df2d506253334290eec9e8aed10c6e287a16d3d9c7fbfe2bbb5eef9a75307ec54002e5efdf6604d657837b19f27995ec073ae2137f8da408c69917ebdb00495686d6d57c4586df33fbe06610663533843509905dd6f3f7d8cb494e8c280b10d1b5725c7062574afbc59338712ad14fe0702e1b4c86f78c639399126f5602040b94a83309647dba95fa1f0a47aa43b62b42dcef062fc9957d22b3e456a04650fdb316d9fa7f973aac9e177142fc47c111d81d82202027265e503f040283641eafa453d1b8397af8f9d69f02c130aa135994f6d916cc' --output-document 'archive.zip'

--2022-06-17 08:16:11--  https://storage.googleapis.com/kaggle-data-sets/1312443/2186313/bundle/archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20220617%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20220617T080425Z&X-Goog-Expires=259199&X-Goog-SignedHeaders=host&X-Goog-Signature=2c284e453351f7bdecda66f4696b76039d2ed56afbab1d23e570ba86533e41acdb5c08d835a8174de45406df2d506253334290eec9e8aed10c6e287a16d3d9c7fbfe2bbb5eef9a75307ec54002e5efdf6604d657837b19f27995ec073ae2137f8da408c69917ebdb00495686d6d57c4586df33fbe06610663533843509905dd6f3f7d8cb494e8c280b10d1b5725c7062574afbc59338712ad14fe0702e1b4c86f78c639399126f5602040b94a83309647dba95fa1f0a47aa43b62b42dcef062fc9957d22b3e456a04650fdb316d9fa7f973aac9e177142fc47c111d81d82202027265e503f040283641eafa453d1b8397af8f9d69f02c130aa135994f6d916cc
Resolving storage.googleapis.com (storage.googleapis.com)... 74.125.200.128, 74.125.68.128, 74.125.24.128, ...
Connecting to storage.g

In [2]:
!unzip archive.zip

Archive:  archive.zip
  inflating: sentiment_tweets3.csv   


In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings
from tqdm import tqdm
import re
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import GridSearchCV
import nltk
from nltk.corpus import stopwords 
from nltk.stem.wordnet import WordNetLemmatizer
import string 
import spacy
import os
import numpy as np

warnings.filterwarnings("ignore")

In [4]:
df=pd.read_csv('sentiment_tweets3.csv')
df.head()

Index                                 message to examine  \
0    106  just had a real good moment. i missssssssss hi...   
1    217         is reading manga  http://plurk.com/p/mzp1e   
2    220  @comeagainjen http://twitpic.com/2y2lx - http:...   
3    288  @lapcat Need to send 'em to my accountant tomo...   
4    540      ADD ME ON MYSPACE!!!  myspace.com/LookThunder   

   label (depression result)  
0                          0  
1                          0  
2                          0  
3                          0  
4                          0

In [5]:
df = df.rename(columns=({'message to examine':'tweet'}))
df = df.rename(columns=({'label (depression result)':'lab'}))
df.columns

Index(['Index', 'tweet', 'lab'], dtype='object')

In [6]:
df.shape
df['lab'].value_counts()

0    8000
1    2314
Name: lab, dtype: int64

looks like data is imbalanced and sorted 

###Text Preprocessing

In [7]:
def cleanhtml(phrase):
    text=re.compile('<.*?>')
    phrase=re.sub(text,' ',phrase)
    return phrase

def decont(phrase):
    # specific
    phrase = re.sub(r"won't", "will not", phrase)
    phrase = re.sub(r"can\'t", "can not", phrase)
    phrase = re.sub(r"PRON", " ", phrase)
    phrase = re.sub(r"pron", " ", phrase)
    #remove any url
    phrase = re.sub(r"http[s]?://\S+"," ", phrase)
    # general
    phrase = re.sub(r"n\'t", " not", phrase)
    phrase = re.sub(r"\'re", " are", phrase)
    phrase = re.sub(r"\'s", " is", phrase)
    phrase = re.sub(r"\'d", " would", phrase)
    phrase = re.sub(r"\'ll", " will", phrase)
    phrase = re.sub(r"\'t", " not", phrase)
    phrase = re.sub(r"\'ve", " have", phrase)
    phrase = re.sub(r"\'m", " am", phrase)

    #special char
    phrase = re.sub(r'[^A-Za-z0-9]+', " ", phrase)

    phrase = re.sub(r"@+"," ", phrase)

    #remove any thing with html tags
    phrase = cleanhtml(phrase)
    return phrase


In [8]:
nltk.download('omw-1.4')
nltk.download('stopwords')
nltk.download('wordnet')
nlp = spacy.load('en_core_web_sm')
lmtzr = WordNetLemmatizer()
def text_preprocess(text):
    lm = []
    text = nlp(text)
    for token in text:
        k=(token.lemma_)
        k=decont(k)
        k=lmtzr.lemmatize(k)
        lm.append(k)
    text = " ".join(lm)
    text = text.translate(str.maketrans("", "", string.punctuation))
    #text = [word for word in text.split() ]#if word.lower() not in stopwords.words('english')]
    text = [word for word in text.split() if word.lower() not in stopwords.words('english')]
    #print(text)
    return " ".join(text)

[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Unzipping corpora/omw-1.4.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.


In [9]:
df['ptt'] = df['tweet'].apply(text_preprocess)

In [10]:
df.head()

Index                                              tweet  lab  \
0    106  just had a real good moment. i missssssssss hi...    0   
1    217         is reading manga  http://plurk.com/p/mzp1e    0   
2    220  @comeagainjen http://twitpic.com/2y2lx - http:...    0   
3    288  @lapcat Need to send 'em to my accountant tomo...    0   
4    540      ADD ME ON MYSPACE!!!  myspace.com/LookThunder    0   

                                                 ptt  
0                         real good moment miss much  
1                                         read manga  
2                                       comeagainjen  
3  lapcat need send accountant tomorrow oddly eve...  
4                add MYSPACE myspace com lookthunder

In [11]:
df.to_csv("tweetp",index=False)

**try glove**

In [12]:
!wget --header 'Host: downloads.cs.stanford.edu' --user-agent 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:100.0) Gecko/20100101 Firefox/100.0' --header 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8' --header 'Accept-Language: en-US,en;q=0.5' --referer 'https://colab.research.google.com/' --header 'DNT: 1' --header 'Upgrade-Insecure-Requests: 1' 'http://downloads.cs.stanford.edu/nlp/data/glove.6B.zip' --output-document 'glove.6B.zip'

--2022-06-17 08:17:57--  http://downloads.cs.stanford.edu/nlp/data/glove.6B.zip
Resolving downloads.cs.stanford.edu (downloads.cs.stanford.edu)... 171.64.64.22
Connecting to downloads.cs.stanford.edu (downloads.cs.stanford.edu)|171.64.64.22|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 862182613 (822M) [application/zip]
Saving to: ‘glove.6B.zip’

glove.6B.zip        100%[===================>] 822.24M  5.25MB/s    in 2m 42s  

2022-06-17 08:20:39 (5.09 MB/s) - ‘glove.6B.zip’ saved [862182613/862182613]



In [13]:
!unzip glove.6B.zip

Archive:  glove.6B.zip
  inflating: glove.6B.50d.txt        
  inflating: glove.6B.100d.txt       
  inflating: glove.6B.200d.txt       
  inflating: glove.6B.300d.txt       


In [14]:
# Create Empty dictionary
word2vector = {}
#Create a dictionary with word and corresponding vector
with open(os.path.join('./glove.6B.100d.txt')) as file:
  for line in file:
    list_of_values = line.split()    	
    word = list_of_values[0]
    vector_of_word = np.asarray(list_of_values[1:], dtype='float32')
    word2vector[word] = vector_of_word

In [15]:
i=0;list_of_sentance=[]
for sentance in df['ptt']:
    list_of_sentance.append(sentance.split())

In [16]:
from tqdm import tqdm
avg_w2v_vectors = []; # the avg-w2v for each sentence/review is stored in this list
for sentence in tqdm(list_of_sentance): # for each review/sentence
    vector = np.zeros(100) # as word vectors are of zero length
    cnt_words =0; # num of words with a valid vector in the sentence/review
    for word in sentence: # for each word in a review/sentence
        try:
            vector += word2vector[word]
            cnt_words += 1
        except:
          continue
    if cnt_words != 0:
        vector /= cnt_words
    avg_w2v_vectors.append(vector)
print(len(avg_w2v_vectors))
print(len(avg_w2v_vectors[0]))

100%|██████████| 10314/10314 [00:00<00:00, 53359.78it/s]

10314
100


In [17]:
X_train, X_test, y_train,y_test = train_test_split(avg_w2v_vectors, df['lab'], test_size=0.25)

In [18]:
X_train=pd.DataFrame(X_train)
X_test=pd.DataFrame(X_test)

In [19]:
xgb_cl = xgb.XGBClassifier(scale_pos_weight=4,tree_method='gpu_hist', gpu_id=0)
clf= SGDClassifier(class_weight={0:4})

In [20]:
param={"alpha":np.logspace(-10,10,11), "penalty":["l1","l2"], "loss":["hinge","log"]}
logreg_cv=GridSearchCV(clf,param,cv=5)
logreg_cv.fit(X_train,y_train)
print("tuned hpyerparameters :(best parameters) ",logreg_cv.best_estimator_)

tuned hpyerparameters :(best parameters)  SGDClassifier(class_weight={0: 4}, penalty='l1')


In [35]:
parameters = {'learning_rate': [0.01,0.02],'n_estimators' : [400,800,1000],'max_depth' : [3,4]}
xgboostcv=GridSearchCV(estimator=xgb_cl, param_grid=parameters, n_jobs = 10, cv = 2, verbose=True)
xgboostcv.fit(X_train, y_train)
xgboostcv.best_estimator_

Fitting 2 folds for each of 12 candidates, totalling 24 fits


XGBClassifier(gpu_id=0, learning_rate=0.02, max_depth=4, n_estimators=1000,
              scale_pos_weight=4, tree_method='gpu_hist')

In [36]:
clf=logreg_cv.best_estimator_
xgb=xgboostcv.best_estimator_

In [37]:
xgb.fit(X_train, y_train)
clf.fit(X_train, y_train)

SGDClassifier(class_weight={0: 4}, penalty='l1')

In [38]:
from sklearn.metrics import confusion_matrix, accuracy_score,roc_auc_score

def conf_matrix_acc(y_true, y_pred):
    print(f'Confusion matrix\n:{confusion_matrix(y_true, y_pred)}\n')
    print(f'Accuracy score is : {accuracy_score(y_true, y_pred)}')

In [39]:
y_pred_xgb = xgb.predict(X_test)
y_pred_sgd = clf.predict(X_test)

In [40]:
conf_matrix_acc(y_test,y_pred_xgb ),print(roc_auc_score(y_test,y_pred_xgb))
print("="*100)
print("="*100)
conf_matrix_acc(y_test,y_pred_sgd ),print(roc_auc_score(y_test,y_pred_sgd ))

Confusion matrix
:[[1988   26]
 [  39  526]]

Accuracy score is : 0.9747964327258627
0.9590319093777186
Confusion matrix
:[[2002   12]
 [ 170  395]]

Accuracy score is : 0.9294300116324157
0.8465783761457409


(None, None)

## We can here by obtain a good accuracy of 0.975 and auc_roc of 0.96 with XGBoost